airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartition.kt

```diff
@@ -0,0 +1,320 @@
+import io.airbyte.cdk.read.WhereClauseLeafNode
+import io.airbyte.cdk.read.WhereClauseNode
+import io.airbyte.cdk.read.optimize
+import io.airbyte.cdk.util.Jsons
+
+/** Base class for default implementations of [JdbcPartition]. */
+sealed class MysqlJdbcPartition(
+    val selectQueryGenerator: SelectQueryGenerator,
+    final override val streamState: DefaultJdbcStreamState,
+) : JdbcPartition<DefaultJdbcStreamState> {
+    val stream: Stream = streamState.stream
+    val from = From(stream.name, stream.namespace)
+}
+
+/** Base class for default implementations of [JdbcPartition] for unsplittable partitions. */
+sealed class MysqlJdbcUnsplittablePartition(
+    selectQueryGenerator: SelectQueryGenerator,
+    streamState: DefaultJdbcStreamState,
+) : MysqlJdbcPartition(selectQueryGenerator, streamState) {
+
+    override val nonResumableQuery: SelectQuery
+        get() = selectQueryGenerator.generate(nonResumableQuerySpec.optimize())
+
+    val nonResumableQuerySpec = SelectQuerySpec(SelectColumns(stream.fields), from)
+
+    override fun samplingQuery(sampleRateInvPow2: Int): SelectQuery {
+        val sampleSize: Int = streamState.sharedState.maxSampleSize
+        val querySpec =
+            SelectQuerySpec(
+                SelectColumns(stream.fields),
+                FromSample(stream.name, stream.namespace, sampleRateInvPow2, sampleSize),
+            )
+        return selectQueryGenerator.generate(querySpec.optimize())
+    }
+}
+
+/** Default implementation of a [JdbcPartition] for an unsplittable snapshot partition. */
+class MysqlJdbcUnsplittableSnapshotPartition(
+    selectQueryGenerator: SelectQueryGenerator,
+    streamState: DefaultJdbcStreamState,
+) : MysqlJdbcUnsplittablePartition(selectQueryGenerator, streamState) {
+
+    override val completeState: OpaqueStateValue = MysqlJdbcStreamStateValue.snapshotCompleted
+}
+
+/**
+ * Default implementation of a [JdbcPartition] for an unsplittable snapshot partition preceding a
+ * cursor-based incremental sync.
+ */
+class MysqlJdbcUnsplittableSnapshotWithCursorPartition(
+    selectQueryGenerator: SelectQueryGenerator,
+    streamState: DefaultJdbcStreamState,
+    val cursor: Field,
+) :
+    MysqlJdbcUnsplittablePartition(selectQueryGenerator, streamState),
+    JdbcCursorPartition<DefaultJdbcStreamState> {
+
+    override val completeState: OpaqueStateValue
+        get() =
+            MysqlJdbcStreamStateValue.cursorIncrementalCheckpoint(
+                cursor,
+                cursorCheckpoint = streamState.cursorUpperBound!!,
+                streamState.stream,
+            )
+
+    override val cursorUpperBoundQuery: SelectQuery
+        get() = selectQueryGenerator.generate(cursorUpperBoundQuerySpec.optimize())
+
+    val cursorUpperBoundQuerySpec = SelectQuerySpec(SelectColumnMaxValue(cursor), from)
+}
+
+/** Base class for default implementations of [JdbcPartition] for splittable partitions. */
+sealed class MysqlJdbcSplittablePartition(
```

Comment:
https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/bulk/toolkits/extract-jdbc/src/main/kotlin/io/airbyte/cdk/read/JdbcPartitionsCreator.kt#L158

I thought splittable here just means resumable or not resumable because of the code above. It sounds like if not splittable the reader won't produce checkpoint state message - means we can't resume from a failed incremental read. Is that by design?